src/main/java/com/google/devtools/build/lib/syntax/Dict.java

```diff
@@ -535,9 +539,10 @@ public boolean containsKey(StarlarkSemantics semantics, Object key) throws EvalE
       Dict<? extends K, ? extends V> left,
       Dict<? extends K, ? extends V> right,
       @Nullable Mutability mu) {
-    Dict<K, V> result = Dict.of(mu);
-    result.putAllUnsafe(left);
-    result.putAllUnsafe(right);
+    Dict<K, V> result = new Dict<>(mu, Math.max(left.size(), right.size()));
```

Comment:
fyi, a linter at Google recommends adding `import static java.lang.Math.max;`
I've applied the suggestion in your commit.